src/proj.rs

```diff
@@ -0,0 +1,791 @@
+    <P as Pixel>::Subpixel: Send + Sync,
+    <P as Pixel>::Subpixel: ValueInto<f32> + Clamp<f32>,
+{
+    let (width, height) = image.dimensions();
+    let mut out = ImageBuffer::new(width, height);
+    warp_with(mapping, image, interpolation, default, &mut out);
+
+    out
+}
+
+/// Warp image with custom function.
+/// This enables the user to define a custom mapping such as a wave pattern:
+/// ```
+/// use image::{ImageBuffer, Luma};
+/// use imageproc::utils::gray_bench_image;
+/// use imageproc::proj::*;
+///
+/// let img = gray_bench_image(300, 300);
+/// let mut out = ImageBuffer::new(300, 300);
+/// warp_with(|x,y| (x, y+(x/30.0).sin()), &img, Interpolation::Nearest, Luma([0u8]), &mut
+/// out);
+/// ```
+pub fn warp_with<P, Fm>(mapping: Fm,
+                   image: &Image<P>,
+                   interpolation: Interpolation,
+                   default: P,
+                   out: &mut Image<P>
+                  )
+    where
+    Fm: Fn(f32, f32) -> (f32, f32) + Send + Sync,
+    P: Pixel + Send + Sync + 'static,
+    <P as Pixel>::Subpixel: Send + Sync,
+    <P as Pixel>::Subpixel: ValueInto<f32> + Clamp<f32>,
+
+{
+    let nn = |x,y| nearest(image, x, y, default);
+    let bl = |x,y| interpolate(image, x, y, default);
+    use Interpolation as I;
+
+    match interpolation {
+        I::Nearest => warp_inner(out, mapping, nn),
+        I::Bilinear => warp_inner(out, mapping, bl),
+    }
+}
+
+fn warp_inner<P,Fc,Fi>(
+    out: &mut Image<P>,
+    coord_tf: Fc,
+    get_pixel: Fi,
+    )
+where
+    P: Pixel + 'static,
+    <P as Pixel>::Subpixel: Send + Sync,
+    <P as Pixel>::Subpixel: ValueInto<f32> + Clamp<f32>,
+    Fc: Fn(f32, f32) -> (f32, f32) + Send + Sync,
+    Fi: Fn(f32, f32) -> P + Send + Sync,
+{
+    let width = out.width();
+    let raw_out = out.as_mut();
+    let pitch = P::CHANNEL_COUNT as usize * width as usize;
+
+    raw_out.par_chunks_mut(pitch).enumerate().for_each(|(y, row)| { 
+        for (x, slice) in row.chunks_mut(P::CHANNEL_COUNT as usize).enumerate() {
+            let (px, py) = coord_tf(x as f32, y as f32);
+            *P::from_slice_mut(slice) = get_pixel(px, py);
+        }
+    });
+        
+}
+
+fn class_from_matrix(mx: [f32; 9]) -> TransformationClass {
+
+    if (mx[6]-0.0).abs() < 1e-10 &&
+       (mx[7]-0.0).abs() < 1e-10 &&
+       (mx[8]-1.0).abs() < 1e-10
+    {  
+        if (mx[0]-1.0).abs() < 1e-10 &&
+           (mx[1]-0.0).abs() < 1e-10 &&
+           (mx[3]-0.0).abs() < 1e-10 &&
+           (mx[4]-1.0).abs() < 1e-10 
+        {
+            TransformationClass::Translation 
+        } else { 
+            TransformationClass::Affine
+        }
+    } else {
+        TransformationClass::Projection
+    }
+}
+
+fn normalize(mx: [f32; 9]) -> [f32; 9] {
+    [mx[0]/mx[8], mx[1]/mx[8], mx[2]/mx[8],
+     mx[3]/mx[8], mx[4]/mx[8], mx[5]/mx[8],
+     mx[6]/mx[8], mx[7]/mx[8], mx[8]/mx[8]]
+}
+
+fn try_inverse(t: &[f32; 9]) -> Option<[f32; 9]> {
+    let (
```

Comment:
This is copy-pasted from the equivalent function from `affine`, with some minor tweaks. Please extract most of the logic into a common function that they can share.